server/src/main/java/org/apache/druid/catalog/model/TableDefn.java

```diff
@@ -0,0 +1,176 @@
+            ),
+            properties
+        )
+    );
+    this.columnDefns = columnDefns == null ? Collections.emptyMap() : toColumnMap(columnDefns);
+  }
+
+  public static Map<String, ColumnDefn> toColumnMap(final List<ColumnDefn> colTypes)
+  {
+    ImmutableMap.Builder<String, ColumnDefn> builder = ImmutableMap.builder();
+    for (ColumnDefn colType : colTypes) {
+      builder.put(colType.typeValue(), colType);
+    }
+    return builder.build();
+  }
+
+  /**
+   * Validate a table spec using the table, field and column definitions defined
+   * here. The column definitions validate the type of each property value using
+   * the object mapper.
+   */
+  public void validate(ResolvedTable table)
+  {
+    validate(table.properties(), table.jsonMapper());
+    validateColumns(table.spec().columns(), table.jsonMapper());
+  }
+
+  public void validateColumns(List<ColumnSpec> columns, ObjectMapper jsonMapper)
+  {
+    if (columns == null) {
+      return;
+    }
+    Set<String> names = new HashSet<>();
+    for (ColumnSpec colSpec : columns) {
+      if (!names.add(colSpec.name())) {
+        throw new IAE("Duplicate column name: " + colSpec.name());
+      }
+      ColumnDefn.ResolvedColumn resolvedCol = resolveColumn(colSpec);
+      resolvedCol.validate(jsonMapper);
+    }
+  }
+
+  /**
+   * Resolve the column type to produce a composite object that holds
+   * both the definition and the column spec.
+   */
+  public ColumnDefn.ResolvedColumn resolveColumn(ColumnSpec spec)
+  {
+    String type = spec.type();
+    if (Strings.isNullOrEmpty(type)) {
+      throw new IAE("The column type is required.");
+    }
+    ColumnDefn defn = columnDefns.get(type);
+    if (defn == null) {
+      throw new IAE("Column type [%s] is not valid for tables of type [%s].", type, typeValue());
+    }
+    return new ColumnDefn.ResolvedColumn(defn, spec);
+  }
+
+  /**
+   * Merge a table spec with an update. The merge affects both the properties and
+   * the list of columns.
+   */
+  public TableSpec merge(TableSpec spec, TableSpec update, ObjectMapper jsonMapper)
+  {
+    String updateType = update.type();
+    if (updateType != null && !spec.type().equals(updateType)) {
+      throw new IAE("The update type must be null or [%s]", spec.type());
+    }
+    Map<String, Object> revisedProps = mergeProperties(spec.properties(), update.properties());
+    List<ColumnSpec> revisedColumns = mergeColumns(spec.columns(), update.columns());
+    TableSpec revisedSpec = new TableSpec(spec.type(), revisedProps, revisedColumns);
+    validate(new ResolvedTable(this, revisedSpec, jsonMapper));
+    return revisedSpec;
+  }
+
+  /**
+   * Merge the set of columns from an existing spec and an update.
+   * Columns are matched by name. If the column exists, then it is updated. If
+   * the column does not exist, then the new column is appended to the existing
+   * list. This merge operation cannot remove columns or change order.
```

Comment:
What is the reason not to allow removal or changing order?

Reply:
Didn't realize that update here could be just a subset of the columns, maybe I missed that in the java doc.